feat: Add support for resetting BatchExportRuns

tomasfarias opened the pull request:
## Problem

There is currently no way to reset a specific BatchExportRun, even though Temporal supports this (for example, [via the CLI](https://docs.temporal.io/tctl-v1/workflow#reset)). When workflows fail, we need to give users the ability to reset them so that they can re-run them.

<!-- Who are we building for, what are their needs, why is this important? -->

## Changes

Added an API endpoint to support resetting a BatchExportRun. The operation is executed by making an RPC call to Temporal. [Temporal considers this an "advanced operation" 
](https://github.com/temporalio/sdk-python/issues/262#issuecomment-1408674678), which is why we require the raw RPC call.

This API call is routed to the `api/projects/:team_id/batch_exports/:export_id/runs/:export_run_id/reset` endpoint (notice that now we have a path for export runs).

Unfortunately, this requires a migration:
* Without the migration, there is no way for us to tie runs that correspond to the same batch together: we only have a reference to the BatchExport (which is a Schedule in Temporal) but not to a set of batch runs. This would have been possible with the original 3 model (BatchExportSchedule, BatchExport, and BatchExportRun) approach, but this way should be fine too.
* I've added a `workflow_id` that uniquely identifies a set of batch runs: when we retry, the `run_id` will be different, but the `workflow_id` will remain the same. This `workflow_id` is set by temporal and set to batch export id + timestamp.
* I've also added a `run_id` to identify specific runs in temporal. This is required for the reset as the reset is acts on a run, not a workflow.

EDIT: This migration is no longer needed.

I was concerned about the existing export runs not being reset-able as they don't have these ID fields. So, I removed the ID fields and do resets without them. Turns out this is viable as:

1.Temporal can handle resets with only a Workflow ID (which we already know the UUID part as it's the BatchExport.id field).
2. We already know the data_interval_end to construct the final part of the Workflow ID.

Moreover, this PR includes some frontend changes: since now we can have multiple runs per batch period, we need a way to display them. I made the ExportRuns table expandable and added another table with individual runs. Made it clear that the top row corresponds to the "last" run by renaming "Status" to "Last status" (as you may have early runs that fail, with a newer one succeeded).

See a DEMO:

![Kooha-2023-06-22-10-51-35](https://github.com/PostHog/posthog/assets/18740659/9b6709d8-02df-4f65-ac2a-5f4b0c7bf974)


<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

tomasfarias committed:
feat: Add support for resetting BatchExportRuns

tomasfarias committed:
feat: Add frontend to support multiple runs per batch

tomasfarias committed:
chore: Bump migration forward

tomasfarias commented:
@hazzadous I'm working through some frontend typing issues, but for the most part this is ready for review 👍 Would love your insights.

tomasfarias commented:
I've made an offering to the type checking gods, but I hated myself while doing it. If any better suggestions are available, I'm happy to take them.

tomasfarias commented:
Ah crap, a new migration was added... Bumping mine again...

tomasfarias committed:
chore: Bump migration to 0332

hazzadous commented on posthog/batch_exports/models.py:
```diff
@@ -64,6 +64,8 @@ class Status(models.TextChoices):
     batch_export = models.ForeignKey(
         "BatchExport", on_delete=models.CASCADE, help_text="The BatchExport this run belongs to."
     )
+    workflow_id = models.TextField(null=True, help_text="The Workflow id as set by Temporal.")
```
There's something that irks me having the temporal implementation bleed over into this part of the data model but I don't have a better solution.

tomasfarias replied:
I removed this.

tomasfarias replied:
I was concerned about the existing export runs not being reset-able as they don't have these ID fields. So, I removed the ID fields and do resets without them. Turns out this is viable as:

1. Temporal can handle resets with only a Workflow ID (which we already know the UUID part as it's the BatchExport.id field).
2. We already know the data_interval_end to construct the final part of the Workflow ID.

hazzadous commented on posthog/batch_exports/service.py:
```diff
@@ -248,21 +251,45 @@ def create_batch_export(team_id: int, interval: str, name: str, destination_data
+    Returns:
+        The run_id assigned to the new run.
+    """
+    request = ResetWorkflowExecutionRequest(
```
What is the advantage of resetting over simply creating a new one via execute_workflow, given that we are resetting to the beginning?

tomasfarias replied:
`execute_workflow` creates a new workflow, with a new workflow ID. So, misusing that could cause duplicate runs. Of course, we could still ensure that ourselves, but reset handles all of that for us. There isn't much of a difference between cancelling + executing vs reset, besides reset being 1 atomic operation.

hazzadous commented on frontend/src/scenes/exports/ViewExport.tsx:
```diff
@@ -177,60 +226,175 @@ const ExportRuns = ({ exportId }: { exportId: string }): JSX.Element => {
+                                        onClick={() => {
+                                            resetExportRun()
+                                                .then(() => {
+                                                    updateCallback(undefined, numberOfRuns)
```
Not important but I wonder if we shouldn't be currying these params into the callback rather than needing to explicitly pass them in each time.

hazzadous reviewed (approved):
Looks great, no blocking comments

tomasfarias committed:
refactor: Remove batch export run temporal id fields

This is so we don't tie ourselves to the temporal implementation. The
ID fields are not really needed for two reasons:
1. Temporal can handle resets with only a Workflow ID (which we
already know the UUID part as it's the BatchExport.id field).
2. We already know the data_interval_end to construct the final part
of the Workflow ID.

NOTE: Currently, this only works for scheduled workflows.
